### ISSUE TITLE
Upgrade to Markdown 2.5.2 and Pygments 2.0.1

### ISSUE THREAD

**Issue by facelessuser** [revolunet/sublimetext-markdown-preview#262]
This issue is to track potential issues or major changes while upgrading.  Work has already begun, and hopefully it won't take too long to get it pushed out.  The upgrade is good because a number of bugs will get fixed, but not everything may be backwards compatible.  There are general changes in motion in the markdown library, and it may require changes to Markdown Preview's configuration if not now, then in the future.  A couple of things to note are:

1. Python 2.6 support has been removed from 2.5.2, so that needs to be added back in for ST2 support.  I am hoping it won't be too much work.

2. Loading of markdown extensions is a bit different, but I think it technically still supports the old way (not sure for how long; and this needs to be confirmed).  This may impact how Markdown Preview allows user configuration of the plugins.  So backwards compatibility may get broken, not sure yet.

3.  Pygments upgrade should hopefully not have any negative effects; most changes should be positive improvements of syntax highlighters and adding support for new languages.  Upgrade will be a little painful as there was originally a number of internal changes to allow for ST2 and ST3 to work with it in the plugin environment.  But overall, it shouldn't have any unexpected surprises.

**Comment by facelessuser** [comment]
It looks like I will be able to preserve backwards compatibility.  But it means not all configurations can be set via Markdown Preview.

### Extension Settings
Before extension settings were set inline when defining what extensions to use:

```
codehilite(guess_lang=False)
```

But now more complex extension settings are allowed, and to set them, you have to feed in a separate variable to Python Markdown, and this cannot be represented in an inline form:

```
extensionConfigs = {
    'smarty': {
        'substitutions': {
            'left-single-quote': '&sbquo;', # sb is not a typo!
            'right-single-quote': '&lsquo;',
            'left-double-quote': '&bdquo;',
            'right-double-quote': '&ldquo;'
        }
    }
}
```
Ideally this would be exposed in Markdown Preview, but that would require a bit of internal rework.  I haven't decided if I will do this yet.  If I did, I would probably require the new way to be used only as it will simplify the current code were me manually handle setting some of this stuff.  I will have to see.

### Extension Loading
Python Markdown no longer allows internal extensions be loaded by just giving there name:

```
codehilite
```

But all extensions are now re required to provide the full package path:

```
markdown.extensions.codehilite
```

Since Markdown Preview has an overload function to ensure extension loading works in the ST2 and ST3 environment, we actually control this.  For this reason, backwards compatibility is currently maintained as I have no intentions of changing this right now.  So Python Markdown has had the `extra` extension modified to support the old way of loading extensions.

**Comment by revolunet** [comment]
hey, nice upgrade @facelessuser, thanks ! 
do you know of any existing helper library to help with the ST2/ST3 compatibility ?

**Comment by facelessuser** [comment]
I have seen small little snippets to manage things like ST2 and ST3 edits, but nothing comprehensive.  On my own plugins, I usually push differences down in an ST API layer and then build common stuff on top of that.  But usually the differences I run into very so greatly between plugins, I haven't bothered to create a library.  ST2 and ST3 differences are usually only about 5% of the issues I run into.  Most are Python2 vs Python3 issues; unicode differences, different libraries etc.

The thing I find most frustrating with Markdown Preview ST2/ST3 compatibility is more the environment itself because ST2 doesn't play nice with the way Pygments and Python Markdown import things.  ST3's environment is quite sane, while ST2 is a huge pain.  Pygments changed some importing in 2.0.1 again, and I had to spend a bit of time figuring out how to get things working again in ST2.

To be honest, I have discontinued support for ST2 on all my plugins as it was just to much work for all the plugins I maintain.

**Comment by revolunet** [comment]
Yep i understand, especially the import thing is very annoying.
Would be useful to have some metrics about ST2/ST3 plugin usage. I'll try to get some data from @wbond :)

**Comment by facelessuser** [comment]
Pygments was always going to be a little annoying due to the way it was designed; it wasn't designed to be portable.  That is why I had to modify it to look at paths in a relative manner, and then modify how it then loads it's lexer, formatter, and style plugins.

The ideal way is to offer a separate plugin that adds Pygments to the system path for all plugins to use, and then convince everyone to rely on that opposed to their own different versions that they add by modifying the paths.  This mainly works best for ST3.  It is a bit more troublesome with ST2.

I basically go through the trouble of making Pygments portable on this plugin to avoid conflicts with other plugins that add various versions of Pygments to the system path.

**Comment by wbond** [comment]
I've only skimmed this issue because of being mentioned. Sounds like perhaps the dependencies functionality of PC 3.0 might be useful? PC 2 and 1 will be EOLed soon, so it should be semi reasonable to rely on PC 3.0. The major downside would be manual installs.

I will see if I can run a query for the install stats over the past 3 months.

**Comment by facelessuser** [comment]
I think PC 3.0 dependencies might be useful.  You still run into the problem of various other plugins all mucking with the system path adding different versions of Pygments etc.  I guess you could just offer the Pygments package as a dependency and the Markdown Preview package.  And then if someone installs another plugin that overrides the Pygments version and creates a problem, you can just tell them to uninstall that offending plugin, or have them convince that plugin's author to move to the provided global Pygments package to resolve conflicts.

The idea is to find a way you can include Pygments in a native (or near native way) so that people don't have to extensively modify the code base to have a non-conflicting version of the plugin; there would be just one version.  And also to provide it globally so people can stop adding multiple different versions to the system path that causes conflicts.  That way you know you are using version X.X.X  which has feature X, Y, and Z when you expect it.

I guess this could pose a problem as upgrading a dependency may potentially break multiple plugins until all the authors can upgrade their respective plugins to work with the new version.  I guess managing your own (in a considerate way as we are doing in this plugin) avoids this kind of issue as you completely control the version you are using in such a way you won't accidentally use someone else's version.  But for libraries that require a great deal of modification to make it portable, this can cause a lot of extra work during upgrades.

Anyways, there are pros and cons either way you go.  Our current method is more labor intensive when upgrading the dependancies and probably more annoying to maintain, but we probably won't break anyone else's plugin when we upgrade, and we probably don't have to worry about someone overriding our expected version.

I don't know yet how well this all would work ST2 vs ST3 (as ST2's environment kind of sucks), but I am fairly confident it would work well for ST3.  Pygments does some interesting things to provide its internal modules, but I optimistically think it would work overall pretty well in this way.

I'm not sure what the really sure what the best solution is. I also need to look into PC 3.0's dependency feature to better understand exactly what is offering.  For now, I will move forward with the pull request as is, but all of this is good to consider for the future.

> The major downside would be manual installs.

I personally just tell anyone who installs manually and has issues to PC.  If you are a developer that wants to contribute to a plugin, you usually know how to install a package manually.  You can just provide a list of dependencies, and they can figure it out.  I don't know...that's just my approach :smile:.

**Comment by facelessuser** [comment]
Resolved in [`1.2.0`](https://github.com/revolunet/sublimetext-markdown-preview/releases/tag/1.2.0)
